### DISCUSSION TITLE
Assigning permissions to a role can not work when using multiple guards

### DISCUSSION THREAD

**Discussion by carveybunt** [spatie/laravel-permission#2479] [Q&A]
```php
        $sysadmin = $role->where('name','system admin')->where('guard_name',BackGuardName())->first();
        $defaultPermissionForSystemAdmin = [
            'study','studyObject','studyObjectOverall',
            'administration','administrationOrganization'
        ];
        $sysadmin->syncPermissions($defaultPermissionForSystemAdmin);
```

Error following:

```
"message": "SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`xuehai_databases`.`role_has_permissions`, CONSTRAINT `role_has_permissions_permission_id_foreign` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE) (Connection: mysql, SQL: insert into `role_has_permissions` (`permission_id`, `role_id`) values (99, 99eb070c-5e0a-4ba2-bd19-38f2267a586e))",
    "exception": "Illuminate\\Database\\QueryException",
    "file": "/var/www/xuehai-api-new/vendor/laravel/framework/src/Illuminate/Database/Connection.php",
    "line": 795,
```

Why does `permission_id` change from `UUID` to `int`? I used uuid.
All permissions and the admin role are the same guard.

**Comment by parallels999** [comment]
Laravel version? Package version?

Seems like you forgot fixing migration on pivots

**Reply by carveybunt** [reply]
laravel: 10.10
spatie/laravel-permission: 5.10

Other functions are normal, and this is wrong. I don't know where I am wrong.

**Reply by parallels999** [reply]
> I don't know where I am wrong.

> Seems like you forgot fixing migration on pivots

Also V6 has better support for uuid:

-  https://github.com/spatie/laravel-permission/pull/2089

**Reply by carveybunt** [reply]
Where is V6? I can't see, and also I don't think the error is by uuid, because I can use uuid in other functions, for example: assigning a role to a user.

**Reply by drbyte** [reply]
> Where is V6?

While it's not yet officially released, you can access it using `^6.0-dev` in composer.

It's stable insomuch as there are no known bugs or incomplete features, but a few more things may be added to it before final release (sometime this month, I hope).

**Reply by parallels999** [reply]
Here is V6 https://packagist.org/packages/spatie/laravel-permission#dev-main
>I don't think the error is by uuid, because I can use uuid in other functions, for example: assigning a role to a user

If you are not going to follow directions, why ask for help?
It doesn't matter what you think, it doesn't matter that it works on `role -> user`, the problem is on `role -> permissions`, and I'm almost sure you have the migration wrong
If you want a more exact solution, you must upload an example app showing the problem
https://github.com/spatie/laravel-permission/blob/feb301470b7ca09e77cca8d2dec662a89d2770a8/tests/TestCase.php#L215-L217
https://github.com/spatie/laravel-permission/blob/feb301470b7ca09e77cca8d2dec662a89d2770a8/tests/TestCase.php#L224-L226
>Seems like you forgot fixing migration on pivots

**Reply by carveybunt** [reply]
thanks, I will try to use V6. But I wanna use `role->permissions`, how can to make that work?

**Reply by drbyte** [reply]
> thanks, I will try to use V6. But I wanna use `role->permissions`, how can to make that work?

I don't understand your question. Can you give an example of what you mean?

**Reply by carveybunt** [reply]
I enabled `uuid` and `team`, and use multiple guards. I want to get the model of the user role after the user login, and then I assign multiple permissions for the role, rather than assigning permissions to the user.
I hope you can understand my description, thank you very much.

**Reply by erikn69** [reply]
> I enabled `uuid` and `team`

Did you set the team_id on assign?

**Comment by erikn69** [comment]
Show your migration, 
I think you did not follow the documentation correctly
https://github.com/spatie/laravel-permission/blob/main/docs/advanced-usage/uuid.md

**Reply by carveybunt** [reply]
Thank you for your reply, I hope my migration is not wrong. My migration is as follows:
```php
      Schema::create(
            $tableNames['permissions'], function(Blueprint $table) {
            //            $table->bigIncrements('id'); // permission id
                $table->uuid('id')->primary(); // permission uuid
                $table->uuid("parent_id")->nullable();
                $table->string('name', 125);       // For MySQL 8.0 use string('name', 125);
                $table->string('guard_name', 125); // For MySQL 8.0 use string('guard_name', 125);
                $table->string("title"); // my add
                $table->tinyInteger("sort")->default(10); // my add
                $table->timestamps();
                //            $table->unique(['name', 'guard_name']);
                $table->unique(["name", "guard_name", "parent_id"]); // my add
            },
        );
        Schema::create(
            $tableNames['roles'], function(Blueprint $table) use ($teams, $columnNames) {
            //            $table->bigIncrements('id'); // role id
                $table->uuid('id')->primary(); // role uuid
                if ( $teams || config('permission.testing') ) { // permission.testing is a fix for sqlite testing
                    //                $table->unsignedBigInteger($columnNames['team_foreign_key'])->nullable();
                    $table->uuid($columnNames['team_foreign_key'])->nullable();
                    $table->index($columnNames['team_foreign_key'], 'roles_team_foreign_key_index');
                }
                $table->string('name', 125);       // For MySQL 8.0 use string('name', 125);
                $table->string('guard_name', 125); // For MySQL 8.0 use string('guard_name', 125);
                $table->timestamps();
                $table->string("title"); // my add
                $table->tinyInteger("sort")->default(10); // my add
                $table
                    ->foreignUuid("department_id")
                    ->nullable()
                    ->constrained("departments")
                    ->onDelete("restrict");  // my add
                if ( $teams || config('permission.testing') ) {
                    $table->unique([$columnNames['team_foreign_key'], 'name', 'guard_name','department_id']);
                } else {
                    //                $table->unique(['name', 'guard_name']);
                    $table->unique(["name", "guard_name", "department_id"]); // my change
                }
            },
        );
       Schema::create(
            $tableNames['model_has_permissions'], function(Blueprint $table) use ($tableNames, $columnNames, $teams) {
            //            $table->unsignedBigInteger(PermissionRegistrar::$pivotPermission);
                $table->uuid(PermissionRegistrar::$pivotPermission);
                $table->string('model_type');
            //            $table->unsignedBigInteger($columnNames['model_morph_key']);
                $table->uuid($columnNames['model_morph_key']);
                $table->index(
                    [$columnNames['model_morph_key'], 'model_type'], 'model_has_permissions_model_id_model_type_index',
                );
                $table->foreign(PermissionRegistrar::$pivotPermission)
                    ->references('id') // permission id
                    ->on($tableNames['permissions'])
                    ->onDelete('cascade');
                if ( $teams ) {
                    //                $table->unsignedBigInteger($columnNames['team_foreign_key']);
                    $table->uuid($columnNames['team_foreign_key']);
                    $table->index($columnNames['team_foreign_key'], 'model_has_permissions_team_foreign_key_index');
                    $table->primary(
                        [
                            $columnNames['team_foreign_key'], PermissionRegistrar::$pivotPermission,
                            $columnNames['model_morph_key'], 'model_type',
                        ],
                        'model_has_permissions_permission_model_type_primary',
                    );
                } else {
                    $table->primary(
                        [PermissionRegistrar::$pivotPermission, $columnNames['model_morph_key'], 'model_type'],
                        'model_has_permissions_permission_model_type_primary',
                    );
                }
            },
        );
       Schema::create(
            $tableNames['model_has_roles'], function(Blueprint $table) use ($tableNames, $columnNames, $teams) {
            //            $table->unsignedBigInteger(PermissionRegistrar::$pivotRole);
                $table->uuid(PermissionRegistrar::$pivotRole);
                $table->string('model_type');
            //            $table->unsignedBigInteger($columnNames['model_morph_key']);
                $table->uuid($columnNames['model_morph_key']);
                $table->index([$columnNames['model_morph_key'], 'model_type'], 'model_has_roles_model_id_model_type_index');
                $table->foreign(PermissionRegistrar::$pivotRole)
                    ->references('id') // role id
                    ->on($tableNames['roles'])
                    ->onDelete('cascade');
                if ( $teams ) {
                    //                $table->unsignedBigInteger($columnNames['team_foreign_key']);
                    $table->uuid($columnNames['team_foreign_key']);
                    $table->index($columnNames['team_foreign_key'], 'model_has_roles_team_foreign_key_index');
    
                    $table->primary(
                        [
                            $columnNames['team_foreign_key'], PermissionRegistrar::$pivotRole,
                            $columnNames['model_morph_key'], 'model_type',
                        ],
                        'model_has_roles_role_model_type_primary',
                    );
                } else {
                    $table->primary(
                        [PermissionRegistrar::$pivotRole, $columnNames['model_morph_key'], 'model_type'],
                        'model_has_roles_role_model_type_primary',
                    );
                }
            },
        );
       Schema::create(
            $tableNames['role_has_permissions'], function(Blueprint $table) use ($tableNames) {
            //            $table->unsignedBigInteger(PermissionRegistrar::$pivotPermission);
            //            $table->unsignedBigInteger(PermissionRegistrar::$pivotRole);
                $table->uuid(PermissionRegistrar::$pivotPermission);
                $table->uuid(PermissionRegistrar::$pivotRole);
                $table->foreign(PermissionRegistrar::$pivotPermission)
                    ->references('id') // permission id
                    ->on($tableNames['permissions'])
                    ->onDelete('cascade');
                $table->foreign(PermissionRegistrar::$pivotRole)
                    ->references('id') // role id
                    ->on($tableNames['roles'])
                    ->onDelete('cascade');
                $table->primary(
                    [PermissionRegistrar::$pivotPermission, PermissionRegistrar::$pivotRole],
                    'role_has_permissions_permission_id_role_id_primary',
                );
            },
        );
        
```

**Comment by drbyte** [comment]
> Why does permission_id change from UUID to int? I used uuid.

Please post your \App\Models\Permission model. If you're using UUID you need to use the HasUuids trait (or manually implement all the UUID logic.

**Reply by carveybunt** [reply]
I have done that.

**Comment by drbyte** [comment]
>  I want to get the model of the user role after the user login, and then I assign multiple permissions for the role, rather than assigning permissions to the user.

That's a bit unusual. But as long as the user has at least one role already assigned, then yes you can do that.

**Reply by carveybunt** [reply]
I already assigned a role to the user, but can't be assigning permission to the role.

**Comment by drbyte** [comment]
In a fresh new app, using multiple guards, using teams, and using UUIDs work as documented. And the test suite tests all those things.

Therefore the problem is in how you've adapted/modified things further. It could be in your custom models. It could be in your migrations. It could be in the way you're using teams (vs the documentation).

After reviewing both of your open discussions, it's apparent that there's more information needed in order for us to provide further help.

The best way to provide enough information is to post your application code to a github repository so we can inspect it. 
If you're unable to post all of your application code for privacy reasons, then the best way to provide the information is to create a new Laravel application in a new repository and push that repository to github so we can review it there. It should be as simple as needed to demonstrate the actual problem. See further instructions in the documentation at: https://spatie.be/docs/laravel-permission/v5/basic-usage/new-app

**Reply by carveybunt** [reply]
This is my application code with `laravel/passport` and `spatie/laravel-permission`. [permission-test link](https://github.com/carveybunt/permission-test)  
I just use the application as an API, with no views.
Very thanks for your reply to me.

**Reply by parallels999** [reply]
@carveybunt You did not follow the instructions. 

This commit should have the base code
```
git remote add origin git@github.com:YOURUSERNAME/REPONAME.git
git push -u origin main
```
This commit should have your customizations
```
git add .
git commit -m "Explain what your commit is about here"
git push origin main
```
As you sent everything together it is impossible to check what you changed

**Reply by drbyte** [reply]
I have installed the application at https://github.com/carveybunt/permission-test

Which specific API URLs are triggering the error?

**Comment by drbyte** [comment]
Errors in your application:

1. `/config/permission.php` must use your extended role/permission models, because they have the HasUuids trait:
```diff
return [

    'models' => [
-        'permission' => Spatie\Permission\Models\Permission::class,
+       'permission' => App\Models\Permission::class,

-        'role' => Spatie\Permission\Models\Role::class,
+        'role' => App\Models\Role::class,
    ],
```

2. In `helpers.php`, the `accountType()` function uses `user_name` instead of `name` in `default`, so lookups fail.
Same in `User::findForPassport()` ... `user_name` should be `name` to match your User model.

3. Minor: `BackUsers` model already uses `use HasRoles`. It does not also need `HasPermissions`, because HasRoles already includes HasPermissions.

4. Minor English issue: the `BackUsers::organizaitons()` relationship is misspelled. You do not have to fix it, but if you do, I suggest a global find/replace because you have used it in many other files as well.
```diff
-organizaitons
+organizations
```

5. `/config/httpCode.php` is not part of this app, but your `CheckUserLogin` middleware uses it to provide response/error messages.

6. `/routes/backApi` -- if you try to access `manage/v1/user/detail` without being logged in, error `Route [login] not defined.` occurs because there is no routed named 'login' in `app/Middleware/Authenticate::redirectTo()`. Probably should return a JSON status and error code instead of a redirect.

---

I think only # 1 in this list is "important".
